lib/puppet/indirector/node/rest.rb

```diff
@@ -4,4 +4,14 @@
 class Puppet::Node::Rest < Puppet::Indirector::REST
   desc "Get a node via REST. Puppet agent uses this to allow the puppet master
     to override its environment."
+
+  def find(request)
+    super(request) do |response|
+      server_version = response['X-Puppet-Version']
+      if SemanticPuppet::Version.parse(server_version).major < SemanticPuppet::Version.parse(Puppet::PUPPETVERSION).major
+        Puppet.warning "Server version #{server_version} is unsupported, downgrading to PSON"
+        Puppet[:preferred_serialization_format] = 'pson'
```

Comment:
I think attempting to downgrade is the right thing to do - this is a good user experience improvement and if we can do it we should IMO. Just opting for moving this logic somewhere so that no matter what it is we're doing, we're likely to be able to take advantage of that downgrade-ability.